atlasdb-client/src/main/java/com/palantir/atlasdb/keyvalue/impl/InMemoryKeyValueService.java

```diff
@@ -479,43 +486,81 @@ public CheckAndSetCompatibility getCheckAndSetCompatibility() {
-            if (!succeeded) {
-                byte[] actual = table.entries.get(key); // Re-fetch, something may have happened between get and replace
-                throwCheckAndSetException(cell, tableRef, oldValue.get(), actual);
-            }
-        } else {
-            byte[] oldContents = putIfAbsent(table, key, contents);
-            if (oldContents != null) {
-                throwCheckAndSetException(cell, tableRef, null, oldContents);
+        runWithLockForWrites(tableRef, table -> {
+            Cell cell = request.cell();
+            Optional<byte[]> oldValue = request.oldValue();
+            byte[] contents = request.newValue();
+
+            Key key = getKey(table, cell, AtlasDbConstants.TRANSACTION_TS);
+            if (oldValue.isPresent()) {
+                byte[] storedValue = table.get(key);
+                boolean succeeded =
+                        Arrays.equals(storedValue, oldValue.get()) && table.replace(key, storedValue, copyOf(contents));
+                if (!succeeded) {
+                    byte[] actual = table.get(key); // Re-fetch, something may have happened between get and replace
+                    throwCheckAndSetException(cell, tableRef, oldValue.get(), actual);
+                }
+            } else {
+                byte[] oldContents = putIfAbsent(table, key, contents);
+                if (oldContents != null) {
+                    throwCheckAndSetException(cell, tableRef, null, oldContents);
+                }
             }
-        }
+            return null;
+        });
     }
 
+    /**
+     * This operation ensures atomicity by locking down the table. This ensures that there are no conflicts with other
+     * writes, and that reads see a consistent view of the world. The exception are reads that return iterators, which
+     * may see a partially applied multiCAS.
+     */
     @Override
     public void multiCheckAndSet(MultiCheckAndSetRequest multiCheckAndSetRequest) throws MultiCheckAndSetException {
-        throw new UnsupportedOperationException("InMemoryKvs does not support multi-checkAndSet operation!");
+        TableReference tableRef = multiCheckAndSetRequest.tableRef();
+        Table tableEntry = getTableMap(tableRef);
+        Lock lock = tableEntry.lock.writeLock();
+        lock.lock();
+        try {
+            Map<Cell, byte[]> actualValues = new HashMap<>();
+            for (Map.Entry<Cell, byte[]> entry :
+                    multiCheckAndSetRequest.expected().entrySet()) {
+                Cell cell = entry.getKey();
+                Key key = getKey(tableEntry, cell, AtlasDbConstants.TRANSACTION_TS);
+                byte[] actual = tableEntry.entries.get(key);
+                actualValues.put(cell, actual);
+                if (!Arrays.equals(actual, entry.getValue())) {
+                    throw new MultiCheckAndSetException(
+                            LoggingArgs.tableRef(tableRef),
```

Comment:
I think we should capture every value where actual was different from expected, as is the behavior in Cassandra Kvs.